File: python_prototype/controllers/pure_pursuit.py
import numpy as np


class PurePursuitController:
    """横向 Pure Pursuit 路径跟踪控制器"""

    def __init__(self, wheelbase=2.7, k=0.1, ld_min=2.0):
        self.L      = wheelbase
        self.k      = k        # 前视距离增益：速度越快看得越远
        self.ld_min = ld_min   # 最小前视距离 (m)

    def compute(self, state, trajectory, min_idx=0):
        """
        state:      [x, y, theta, v]
        trajectory: N×4 数组，每行 [x, y, yaw, v_ref]
        min_idx:    最小搜索起点，防止最近点向后跳
        返回:       (转角 delta rad, 目标速度 v_ref, nearest_idx)
        """
        x, y, theta, v = state

        ld = max(self.k * v + self.ld_min, self.ld_min)

        # 只在 min_idx 之后搜索最近点，防止轨迹弯曲时向后跳
        search_start = max(min_idx, 0)
        dists = np.linalg.norm(trajectory[search_start:, :2] - np.array([x, y]), axis=1)
        nearest_idx = search_start + int(np.argmin(dists))

        # 找前视点（距离 >= ld 的第一个点）
        target_idx = nearest_idx
        for i in range(nearest_idx, len(trajectory)):
            if np.linalg.norm(trajectory[i, :2] - np.array([x, y])) >= ld:
                target_idx = i
                break

        tx, ty = trajectory[target_idx, :2]
        alpha = np.arctan2(ty - y, tx - x) - theta
        alpha = self._normalize_angle(alpha)

        delta = np.arctan2(2 * self.L * np.sin(alpha), ld)
        v_ref = trajectory[target_idx, 3]

        return delta, v_ref, nearest_idx

    @staticmethod
    def _normalize_angle(angle):
        while angle >  np.pi: angle -= 2 * np.pi
        while angle < -np.pi: angle += 2 * np.pi
        return angle

File: cpp_controllers/src/pure_pursuit.cpp
#include "pure_pursuit.hpp"
#include <cmath>
#include <limits>
#include <algorithm>

PurePursuitController::PurePursuitController(
    double wheelbase, double k, double ld_min)
    : L_(wheelbase), k_(k), ld_min_(ld_min) {}

std::pair<double, double> PurePursuitController::compute(
    const Eigen::Vector4d& state,
    const std::vector<Waypoint>& trajectory,
    int min_idx)
{
    const double x     = state[0];
    const double y     = state[1];
    const double theta = state[2];
    const double v     = state[3];

    double ld = std::max(k_ * v + ld_min_, ld_min_);

    // 只在 min_idx 之后搜索，防止向后跳
    int start = std::max(min_idx, 0);
    double min_dist = std::numeric_limits<double>::max();
    int nearest = start;
    for (int i = start; i < (int)trajectory.size(); ++i) {
        double d = std::hypot(trajectory[i].x - x, trajectory[i].y - y);
        if (d < min_dist) { min_dist = d; nearest = i; }
    }
    last_nearest_idx_ = nearest;

    // 从最近点向前找前视点
    int target = nearest;
    for (int i = nearest; i < (int)trajectory.size(); ++i) {
        if (std::hypot(trajectory[i].x - x, trajectory[i].y - y) >= ld) {
            target = i;
            break;
        }
    }

    double alpha = std::atan2(trajectory[target].y - y,
                               trajectory[target].x - x) - theta;
    alpha = normalizeAngle(alpha);

    double delta = std::atan2(2.0 * L_ * std::sin(alpha), ld);

    return {delta, trajectory[target].v_ref};
}

double PurePursuitController::normalizeAngle(double angle) {
    while (angle >  M_PI) angle -= 2.0 * M_PI;
    while (angle < -M_PI) angle += 2.0 * M_PI;
    return angle;
}
